### ISSUE TITLE
gmail_auth() on shiny server

### ISSUE THREAD

**Issue by howard-gu** [jennybc/send-email-with-r#13]
Hi Jenny,

First of all, huge fan of your work. I've created a Shiny app which accepts various inputs, and I'd like to send out an email when a user submits inputs.

I've run it locally and it all works, but having trouble getting gmail_auth() to work on the shiny server.

It comes up with the error:
/usr/bin/xdg-open: 851: /usr/bin/xdg-open: www-browser: not found
/usr/bin/xdg-open: 851: /usr/bin/xdg-open: links2: not found
/usr/bin/xdg-open: 851: /usr/bin/xdg-open: elinks: not found
/usr/bin/xdg-open: 851: /usr/bin/xdg-open: links: not found
/usr/bin/xdg-open: 851: /usr/bin/xdg-open: lynx: not found
/usr/bin/xdg-open: 851: /usr/bin/xdg-open: w3m: not found
xdg-open: no method available for opening <  very long url link  >

Understandably on the linux server it can't open up a web server and get me to login. I've tried to circumvent this by copying the .httr-oauth token I generate locally, onto the shiny server. But when I try to gmail_auth(), it still tries to open up the browser, instead of using the token that's there already.

Any ideas on how I can send emails from the shiny server?

Thanks,
Howard

**Comment by jennybc** [comment]
The `.httr-oauth` file is almost certainly not being found. Traditionally, this is usually a path problem and that's very easy to believe in the Shiny context, possibly with an added wrinkle re: who the app is running as and file permissions. So the first line of debugging would be getting visibility into working directory, what files are there and readable, etc.

For unattended use, one should really use a service account token. I'm actively working on a package for use across multiple Google APIs to make all of this easier. Off the top of my head, I'm not sure how easy it is to take this advice with gmailr at this very moment. But it's probably possible.

**Comment by howard-gu** [comment]
Thanks Jenny - I've checked the path / set the working directory for R to be able to find it, but it seems like for some reason R on Linux doesn't automatically discover hidden files, leading it to not find the .httr-oauth token. Not sure how to solve this but would be keen to hear if you have any solutions.

Alternatively I'm trying to authenticate as you mentioned above, using service account token using the googleAuthR package - but then I can't seem to figure out how to get that to interact with gmailR. After I authenticate using googleAuthR package, gmailR doesn't seem to register than it's already authenticated / logged in. Is there anyway I could bridge between the two packages?

Thanks!

**Comment by jennybc** [comment]
I suspect your best bet is to push on the "not finding hidden files" thing. That doesn't sound quite right to me, i.e. the hypothesis that "R on Linux doesn't automatically discover hidden files". I would continue to push very hard on affirming the working directory and listing the files in it and making darn sure you see `.httr-oauth` in that listing. But if that somehow ends up being the problem, it is possible to take charge of httr's cache file path via the option `httr_oauth_cache`. So you could also try setting that to whatever you want (with no leading dot!) and renaming `.httr-oauth` to that. But this explanation still doesn't smell right to me.

I'm not sure if gmailr is currently set up favourably for service account tokens, so let's regard that as plan B or C.

**Comment by howard-gu** [comment]
Just double checking the option setting.

Is this getOption(gmailr.httr_oauth_cache) or just getOption(httr_oauth_cache)?

I can't seem to see a preset default value.

Thanks for all the help!

**Comment by jennybc** [comment]
No, it's just `httr_oauth_cache`. A default value is set in httr's `.onLoad()`.

https://github.com/r-lib/httr/blob/master/R/zzz.R

At least, I don't know of any special arrangements made in gmailr. I know httr's source much better than gmailr's.

**Comment by howard-gu** [comment]
Hi Jenny,

I have finally managed to get it to work, although the reasons of previous failure escapes me. Previously I had set my local R's working directory to a Google Drive directory (using Google Drive Filestream). For some reason the .httr-oauth token which gets created in the Google Drive directory doesn't work when copied over to the shiny server.

However, a .httr-oauth token created after setting the R working directory to a local path (C:\ ... ) works on Shiny-server.

Thanks for all your help!
